gsoc-2023: Add work product for GSoC'23 (#13)
unikraft/gsoc: gsoc-2023/work-product/Rares-Miculescu-Plat-Re-Arch/Rares-Miculescu-plat-rearch.md

# re:Arch Unikraft

<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-2023-badge.svg" align="right" />

## Summary

> Unikraft is an automated system for building specialized POSIX-compliant OSes.

Reviewer: ```suggestion
Unikraft is an automated system for building specialized POSIX-compliant OSes, as unikernels.
```

These unikernels are configured for the needs of specific applications.

> Being based on the concept that "everything is a library", Unikraft should have some architectural changes that can improve the program’s modularity and everybody’s experience. 

Reviewer: ```suggestion
Being based on the concept that "everything is a library", Unikraft requires architectural changes that can improve the modularity of its code base, together with improved developer and maintainer experience.
```

> This project consists in the reorganization of the architecture, through multiple tasks and subtasks.

Reviewer: ```suggestion
This project consists in the reorganzing of platform and architecture code, through multiple tasks and subtasks.
```

For example, the code is organised in multiple sections.
Three out of five directories contain libraries.
The sections describe the purpose of the libraries in detail.

> Based on conventions and ease of work, some libraries were moved to other sections (drivers) or new ones were created, with code that was not part of a library (e.g. ukintctlr).

Reviewer: ```suggestion
Based on conventions and ease of work, some libraries were moved to other sections (drivers) or new ones were created, with code that was not part of a library (e.g. `ukintctlr`).
```

## GSoC contributor

Name: Rares Miculescu

Email: <miculescur@gmail.com>

Github profile: [rares-miculescu](https://github.com/rares-miculescu)

## Mentors

* [Razvan Deaconescu](https://github.com/razvand)

* [Michalis Pappas](https://github.com/michpappas)

* [Marc Rittinghaus](https://github.com/marcrittinghaus)

* [Simon Kuenzer](https://github.com/skuenzer)

> * [Sergiu Moga](https://github.com/mogasergiu)

Reviewer: ```suggestion
* [Razvan Deaconescu](https://github.com/razvand)
* [Simon Kuenzer](https://github.com/skuenzer)
* [Sergiu Moga](https://github.com/mogasergiu)
* [Michalis Pappas](https://github.com/michpappas)
* [Marc Rittinghaus](https://github.com/marcrittinghaus)
```

order alphabetically

## Contributions

> 

Reviewer: Add a paragraph summarizing your contributions overall.

* [include/uk: Move compiler definitions from essentials.h to compiler.h](https://github.com/unikraft/unikraft/pull/960)

* [lib: Add ofw from plat/drivers to lib](https://github.com/unikraft/unikraft/pull/966)

* [plat: Migrate gic to drivers/ukintctlr](https://github.com/unikraft/unikraft/pull/971)

* [drivers: Moving virtio from plat/drivers/ to drivers/](https://github.com/unikraft/unikraft/pull/967)

* [plat: Migrate rtc pl031 to drivers/ukrtc/](https://github.com/unikraft/unikraft/pull/972)

## Blog posts

You can find my blog posts here:

* [Blog post #1](https://github.com/unikraft/docs/pull/268)

* [Blog post #2](https://github.com/unikraft/docs/pull/289)

* [Blog post #3](https://github.com/unikraft/docs/pull/302)

* [Blog post #4](https://github.com/unikraft/docs/pull/307)

## Current status

Besides creating a pull request with the definition of data types, the work is done.
The pull requests need to be reviewed by the mentors.

## Future work

The project is far from finished, so I will keep working on other tasks that need to be done, alongside other members of Unikraft.

> One task that we were discussing on proceeding after the merge of the pull requests, is the future `ukboot` part of plat re-arch.

Reviewer: Also add "I plan to assist in other platform / architecture-related work, such as the addition of new platforms (VMware / Hyper-V) or new architectures (RISC-V)."

I would like to look into VMware support in the near future, given your recent experience, and it would be good to state that.
